SDIOBlockDevice.cpp

```diff
@@ -0,0 +1,378 @@
+    }
+
+    //uint32_t *buffer = static_cast< uint32_t *>(b);
+    // HAL layer uses uint32_t for addr/size
+    uint32_t *buffer = (uint32_t *)(b);
+
+    // Get block count
+    bd_size_t blockCnt = size / _block_size;
+    addr = addr / _block_size;
+
+    // make sure card is ready
+    while (SD_GetCardState() != SD_TRANSFER_OK)
+    {
+        // wait until SD ready
+        wait_ms(1);
+    }
+
+    int status = SD_WriteBlocks_DMA(buffer, addr, blockCnt);
+    debug_if(SD_DBG, "WriteBlocks dbgtest addr: %lld  blockCnt: %lld \n", addr, blockCnt);
+
+    if (status == MSD_OK)
+    {
+        // wait until DMA finished
+        while (SD_DMA_WritePending() != SD_TRANSFER_OK)
+        {
+            uint32_t tickstart = HAL_GetTick();
+            if ((HAL_GetTick() - tickstart) >= MBED_CONF_SD_TIMEOUT)
+            {
+                unlock();
+                status = SD_BLOCK_DEVICE_ERROR_WRITEBLOCKS;
+            }
+        }
+        // make sure card is ready
+        while (SD_GetCardState() != SD_TRANSFER_OK)
+        {
+            // wait until SD ready
+            wait_ms(1);
+        }
+    }
+    else
+    {
+        debug_if(SD_DBG, "WriteBlocks failed! addr: %lld  blockCnt: %lld \n", addr, blockCnt);
+        debug_if(SD_DBG, "  hsd.errorcode: %lu  0x%lx\n", hsd.ErrorCode, hsd.ErrorCode);
+        status = SD_BLOCK_DEVICE_ERROR_WRITEBLOCKS;
+    }
+
+    unlock();
+    return status;
+}
+
+int SDIOBlockDevice::trim(bd_addr_t addr, bd_size_t size)
+{
+    debug_if(SD_DBG, "trim Card...\r\n");
+    lock();
+    if (isPresent() == false)
+    {
+        unlock();
+        return SD_BLOCK_DEVICE_ERROR_NO_DEVICE;
+    }
+    if (!_is_valid_trim(addr, size))
+    {
+        unlock();
+        return SD_BLOCK_DEVICE_ERROR_PARAMETER;
+    }
+
+    if (!_is_initialized)
+    {
+        unlock();
+        return SD_BLOCK_DEVICE_ERROR_NO_INIT;
+    }
+
+    bd_size_t blockCnt = size / _block_size;
+    addr = addr / _block_size;
+
+    int status = SD_Erase(addr, blockCnt);
+    if (status != 0)
+    {
+        debug_if(SD_DBG, "Erase blocks failed! addr: %lld  blockCnt: %lld \n", addr, blockCnt);
+        status = SD_BLOCK_DEVICE_ERROR_ERASEBLOCKS;
+    }
+    else
+    {
+        while (SD_GetCardState() != SD_TRANSFER_OK)
+        {
+            // wait until SD ready
+            wait_ms(10);
+        }
+    }
+
+    unlock();
+    return status;
+}
+
+bd_size_t SDIOBlockDevice::get_read_size() const
+{
+    return _block_size;
+}
+
+bd_size_t SDIOBlockDevice::get_program_size() const
+{
+    return _block_size;
+}
+
+bd_size_t SDIOBlockDevice::size() const
+{
+    return _block_size * _sectors;
+}
+
+void SDIOBlockDevice::debug(bool dbg)
+{
+}
+
+bool SDIOBlockDevice::_is_valid_trim(bd_addr_t addr, bd_size_t size)
+{
+    return (
+        addr % _erase_size == 0 &&
+        size % _erase_size == 0 &&
+        addr + size <= this->size());
+}
+
+bool SDIOBlockDevice::isPresent(void)
+{
+    if (_cardDetect.is_connected())
```

Comment:
mbed-os style requires {} for single line if conditions as well

Reply:
ok